src/shared/theme/resolve.ts

```diff
@@ -69,34 +90,79 @@ function resolveToken(theme: Theme, token: string, path: Array<string>, state?:
 
   if (isReference(assignment)) {
     const ref = getReference(assignment);
-    return resolveToken(theme, ref, path, state);
+    return resolveToken(theme, ref, path, state, baseTheme);
   } else {
     return assignment;
   }
 }
 
-export function resolveContext(theme: Theme, context: Context): FullResolution {
+export function resolveContext(
+  theme: Theme,
+  context: Context,
+  baseTheme?: Theme,
+  themeResolution?: FullResolution
+): FullResolution {
   const tmp = cloneDeep(theme);
+
+  if (!baseTheme || !themeResolution) {
+    tmp.tokens = {
+      ...tmp.tokens,
+      ...context.tokens,
+    };
+    return resolveTheme(tmp, baseTheme);
+  }
+
+  /**
+   * The precedence of context tokens as specified by the API from highest to lowest is:
+   * [override theme context] > [base theme context] > [override theme] [base theme].
+   *
+   * The precedence of tokens as defined in the generated CSS follows this order.
+   * However, tokens that are declared in both the base theme and base theme
+   * context and share the same value are only included in the base theme css. This
+   * results in override theme tokens incorrectly taking precedence over base theme
+   * context.
+   *
+   * To counteract this we can re-baseline the override context using all keys used
+   * in the override theme with their respective values from the base theme context
+   */
+  const baseContext = baseTheme.contexts[context.id];
   tmp.tokens = {
-    ...tmp.tokens,
+    ...Object.keys(themeResolution).reduce((acc, key) => {
+      const shouldSkipReset =
+        (!(key in baseContext.tokens) && !(key in theme.tokens)) ||
+        areAssignmentsEqual(
+          baseContext.tokens[key],
+          theme.tokens[key] ?? baseTheme.tokens[key] // resolved key may not be in override theme
+        );
+
+      return shouldSkipReset
+        ? acc
+        : {
+            ...acc,
+            [key]: baseContext.tokens[key] ?? theme.tokens[key] ?? baseTheme.tokens[key],
+          };
+    }, {}),
     ...context.tokens,
   };
-  return resolveTheme(tmp);
+
```

Comment:
I debated adding another optimization here, but I questioned whether I should rely on the css optimization that cloudscape does (described in the `resolveContext` code comment) so I ultimately decided to leave it out. But in my testing with the ~45 tokens I was using, it reduced the css size from ~120KB to ~70KB. The gzip value remained the same at 4KB, so this optimization would only mean less css to cache server side but probably wouldn't change TTI, etc. much

```typescript
/**
 * Relying on the "override theme tokens incorrectly taking precedence over base theme
 * context" behavior described above when it is advantageous
 */
Object.keys(context.tokens).forEach((key) => {
  const isDuplicateAssignment =
    areAssignmentsEqual(context.tokens[key], theme.tokens[key]) &&
    (!(key in baseContext.tokens) || areAssignmentsEqual(baseContext.tokens[key], baseTheme.tokens[key]));

  if (isDuplicateAssignment) {
    delete tmp.tokens[key];
  }
});
```

Reply:
We came up with the same outcome last time we looked into that. We can possibly look into refactoring this all together, to not produce duplicate values from the start, but "inflate then clean" pattern is more prone to errors, so let's keep it this way.